api/v1beta1/glance_webhook.go

```diff
@@ -23,19 +23,51 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
+// GlanceDefaults -
+type GlanceDefaults struct {
+	ContainerImageURL string
+}
+
+var glanceDefaults GlanceDefaults
+
 // log is for logging in this package.
 var glancelog = logf.Log.WithName("glance-resource")
 
 // SetupWebhookWithManager sets up the webhook with the Manager
-func (r *Glance) SetupWebhookWithManager(mgr ctrl.Manager) error {
+func (r *Glance) SetupWebhookWithManager(mgr ctrl.Manager, defaults GlanceDefaults) error {
```

Comment:
So when we move the defaulting to run from the openstack-operator we need to make sure that the defaults are provided to `func (spec *GlanceSpec) Default()` in some way. In that case the SetupWebhookWithManager call well have  OpenStackOperator receiver. So maybe we need to change the signature to `func (spec *GlanceSpec) Default(defaults)` to be able to pass in some defaults from openstack-operator to the glance defaulter.

Reply:
I actually had it that way initially.  But then I realized that the env vars are there in the Glance deployment within the OpenStack operator CSV, so the env vars will be acquired and passed to the webhooks when Glance's `main.go` is executed...unless we have the OpenStack operator set `ENABLE_WEBHOOKS=false` on the sub-deployments, which I know we are considering.

Reply:
Yeah I think you are correct. So the only "problematic" case is when the webhook is only enabled in the openstack-operator but not in the glance-operator

Reply:
Just wondering why we want to use env variables for default images and change the webhook setup? As I understand the only things that can change for images are [1] and those can be inputs (like storageClass[2]) and there can be defaults for those (ex. tag: latest and namespace: quay.io/tripleomaster or openshift internal registry namespace). The defaulting logic here or the reconciler code could use those to populate containerImage field.

[1]
namespace: quay.io/tripleozedcentos9 (or what is used by the container build pipeline)
name_prefix: openstack-
tag: current-tripleo

ex. quay.io/tripleozedcentos9/openstack-<mariadb/service_name>:current-tripleo

[2] https://github.com/openstack-k8s-operators/openstack-operator/blob/master/apis/core/v1beta1/openstackcontrolplane_types.go#L47

Reply:
> Just wondering why we want to use env variables for default images and change the webhook setup? As I understand the only things that can change for images are [1] and those can be inputs (like storageClass[2]) and there can be defaults for those (ex. tag: latest and namespace: quay.io/tripleomaster or openshift internal registry namespace). The defaulting logic here or the reconciler code could use those to populate containerImage field.
> 
> [1] namespace: quay.io/tripleozedcentos9 (or what is used by the container build pipeline) name_prefix: openstack- tag: current-tripleo
> 
> ex. quay.io/tripleozedcentos9/openstack-<mariadb/service_name>:current-tripleo
> 
> [2] https://github.com/openstack-k8s-operators/openstack-operator/blob/master/apis/core/v1beta1/openstackcontrolplane_types.go#L47

This seems like something to think about.
Is there more background somewhere on they why for this change?

Reply:
@dprince Could you provide the background here as to why we're repeating this pattern that we also used in OSPdO [1]?

https://github.com/openstack-k8s-operators/osp-director-operator/blob/8006828eaa2c4eff79975b4afc2ee3fdf7936471/main.go#L328-L355

Reply:
@hjensas So the context around why we would want to use env vars to provide defaults for service container images involves CSV, `relatedImages` and offline/air-gapped installations.  See the `IMPORTANT` block here: https://docs.openshift.com/container-platform/4.12/operators/admin/olm-restricted-networks.html